### DISCUSSION TITLE
NER for partially annotated documents

### DISCUSSION THREAD

**Discussion by rverdier65** [explosion/spaCy#12239] [Help: Model Advice] [answered]
Hello ! 

We are using spacy to learn multiple tasks from a documents corpus:

- Global document labelling (textcat_multilabel)
- Document line labelling (spancat)
- Named entity recognition (ner)

We have good results for the two first tasks. But we encounter some difficulties for the ner.

## Problem definition

Let’s assume the entities are ENT1 and ENT2.

Our dataset is **sparse** and some documents are **not fully labelled.** For a given document we can have either:

- ENT1 and ENT2 that has been reviewed and annotated (for example, if there is no annotated entries of ENT2 for this document, it means that it could not have one)
- ENT1 or ENT2 (or both) that has not been reviewed and annotated (for example, if there is no annotated entries of ENT2 for this document, it means that it could have one, but this entity has not been reviewed for this document and we have this info)

We **know** for each document which **entities** has been reviewed and annotated.

We want to **use all documents** to learn to predict every entities.

## Question

- How could we prevent the model to learn not to predict **ENT2** in a document where **ENT2** has not been annotated, whereas it contains tokens corresponding to **ENT2 ?**
**Constraint**: we still want to use such a document to learn to predict **ENT1.**

Thank you very much for your help !

**Comment by rmitsch** [comment]
Hi @rverdier65,

> ENT1 and ENT2 that has been reviewed and annotated (for example, if there is no occurence of ENT2 in the dataset, it means that it could not have one)

You say "For a given document", but refer to a "dataset" in this sentence. Did you mean to say "document", i. e. "if there is no occurence of ENT2 in the document"?

In general, could you elaborate on what you mean with "reviewed"?

**Comment by rverdier65** [comment, marked as answer]
Hi @rmitsch, thank you for your answer.

Indeed I am refering to the document.
I meant to say "if there is no annotated entries of ENT2 for this document"

What I mean with 'reviewed and annotated' is the following: 
For a given entity, for example ENT2, we consider that the document has been 'reviewed and annotated' by a user, if the user have exhaustively annotated all the occurences of ENT2 in the document.
There is 2 possibilities:
- The document contains some tokens corresponding to ENT2, so all of these tokens have been annotated by the user
- The document does not contain any token corresponding to ENT2, so the user just flagued that this document has been reviewed for ENT2 but there is nothing to annotate

So for each document, we know which entity has been 'reviewed and annotated'. Some documents can have only ENT1 'reviewed an annotated', others only ENT2, other both and other none of them.

We want to use all documents, even if they are 'partially annotated' = not all the entities have been 'reviewed an annotated' in the document.

Is it clearer ?

**Reply by rmitsch** [reply]
I think I got it now. To recap, every document is in of three sets:
- Has ENT1 and ENT2 entities
- Has ENT1, but no ENT2 entities
- HAS ENT2, but no ENT1 entities

And you _know_ which document is in which set - also at inference time. And you don't want your NER model to predict entities of the wrong type for a document of which you know it has no such entities.

I'm wondering - what's the issue with your NER model predicting e. g. ENT2 for an ENT1-only document? You can just ignore those predictions after all, as you know that these predictions can't be correct.

**Reply by rverdier65** [reply]
No it is not that. Sorry for the misunderstanding, quite difficult to explain…

To understand better, we can simplify our problem as the following:

Let's assume each document has ENT1 and ENT2 entities.
Every dataset document is in of three sets:
- ENT1 and ENT2 have been annotated
- ENT1 has been annotated but no ENT2
- ENT2 has been annotated but no ENT1

That is why I am speaking about partially annotated documents: some documents are fully annotated, for some documents only ENT1 is annotated, for others only ENT2 is annotated.

We want to use all of these documents to train our NER.

Is it clearer ?

**Reply by rmitsch** [reply]
Ah, I see. Can you elaborate on why you don't want your NER model to predict ENTX entities in documents in which ENTX has not been annotated (yet)?

**Reply by rverdier65** [reply]
If we use the current NER architecture, during training, the model will predict ENTX in documents in which ENTX has not been annotated. So, the model will learn not to predict ENTX on such documents and will not be accurate on similar document.
You see ?

**Reply by rmitsch** [reply]
Yes, I understand the issue - thanks for clarifying. This is not supported by spaCy's NER, as labels are not treated independently: the model learns to predict all specified labels for every document. 

If you don't want that, I'd advise you to train two different models - one for ENT1, one for ENT2. Then train those models only on documents that have been annotated correspondingly.

**Reply by rverdier65** [reply]
Thank you for the advice !
Unfortunatly we have this problem for a lot of entities so we can not afford to train one model by entity.

Don't you have other advices ?

- We tried to patch the first backprop [here](https://github.com/explosion/spaCy/blob/master/spacy/pipeline/transition_parser.pyx#L418), setting to 0 the gradient of the non annotated entities. It improved a lot the results we had but it is not sufficient. And now we have a lot of difficulties to patch this [one](https://github.com/explosion/spaCy/blob/master/spacy/pipeline/transition_parser.pyx#L426).

- We are thinking about testing a new approch, using spancat component with Begin and Inside label for each entity (B-ENT1, I-ENT1, B-ENT2, I-ENT2), and then patch the loss [here](https://github.com/explosion/spaCy/blob/master/spacy/pipeline/spancat.py#L404) which seems easier.

What do you think about this ?

**Reply by rmitsch** [reply]
To back up a bit here, I think this is probably not the best direction to approach this problem from. 

Is it possible for you to complete annotations for your set of documents? Hacking around with gradients is a possibility, but it's untested and runs counter to the design goals for the NER architecture. I'd recommend not do do it, if it's in any way avoidable.

**Reply by rverdier65** [reply]
Thank you for you suggestion but no, it is not possible to complete the annotations because our model need to be able to learn new tasks quickly without annotated all the documents for these new tasks. And also because it would be a looot of documents to annotate.
So, we will always have some entities not annotated on all documents.

We made a poc for the second option and it is very encouraging.

**Reply by rmitsch** [reply]
Glad to hear that your POC is going well :slightly_smiling_face:  Feel free to post more details here if you want to share your findings with rest of the community!

**Reply by rverdier65** [reply]
Thank you for your answers ! 🙂
